pkg/devfile/generator/generators_test.go

```diff
@@ -0,0 +1,570 @@
+			},
+			wantPorts: []corev1.ServicePort{
+				{
+					Name:       "port-8080",
+					Port:       8080,
+					TargetPort: intstr.FromInt(8080),
+				},
+				{
+					Name:       "port-9090",
+					Port:       9090,
+					TargetPort: intstr.FromInt(9090),
+				},
+			},
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+
+			devObj := parser.DevfileObj{
+				Data: &testingutil.TestDevfileData{
+					Components: tt.containerComponents,
+				},
+			}
+
+			serviceSpec, err := GetService(devObj, tt.labels)
+
+			// Unexpected error
+			if (err != nil) != tt.wantErr {
+				t.Errorf("TestGetService() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+
+			// Expected error and got an err
+			if tt.wantErr && err != nil {
+				return
+			}
+
+			if !reflect.DeepEqual(serviceSpec.Selector, tt.labels) {
+				t.Errorf("expected service selector is %v, actual %v", tt.labels, serviceSpec.Selector)
+			}
+			if len(serviceSpec.Ports) != len(tt.wantPorts) {
+				t.Errorf("expected service ports length is %v, actual %v", len(tt.wantPorts), len(serviceSpec.Ports))
+			} else {
+				for i := range serviceSpec.Ports {
+					if serviceSpec.Ports[i].Name != tt.wantPorts[i].Name {
+						t.Errorf("expected name %s, actual name %s", tt.wantPorts[i].Name, serviceSpec.Ports[i].Name)
+					}
+					if serviceSpec.Ports[i].Port != tt.wantPorts[i].Port {
+						t.Errorf("expected port number is %v, actual %v", tt.wantPorts[i].Port, serviceSpec.Ports[i].Port)
+					}
+				}
+			}
+		})
+	}
+
+}
+
+func TestGetBuildConfig(t *testing.T) {
+
+	tests := []struct {
+		name           string
+		GitURL         string
+		GitRef         string
+		ImageName      string
+		ImageNamespace string
+	}{
+		{
+			name:           "Case 1: Get a Source Strategy Build Config",
+			GitURL:         "url",
+			GitRef:         "ref",
+			ImageName:      "image",
+			ImageNamespace: "namespace",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			commonObjectMeta := GetObjectMeta(tt.ImageName, tt.ImageNamespace, nil, nil)
+			buildStrategy := GetSourceBuildStrategy(tt.ImageName, tt.ImageNamespace)
+			params := BuildConfigParams{
+				CommonObjectMeta: commonObjectMeta,
+				BuildStrategy:    buildStrategy,
+				GitURL:           tt.GitURL,
+				GitRef:           tt.GitRef,
+			}
+			buildConfig := GetBuildConfig(params)
+
+			if buildConfig.Name != tt.ImageName {
+				t.Error("TestGetBuildConfig error - build config name does not match")
+			}
+
+			if !strings.Contains(buildConfig.Spec.CommonSpec.Output.To.Name, tt.ImageName) {
+				t.Error("TestGetBuildConfig error - build config output name does not match")
+			}
+
+			if buildConfig.Spec.Source.Git.Ref != tt.GitRef || buildConfig.Spec.Source.Git.URI != tt.GitURL {
+				t.Error("TestGetBuildConfig error - build config git source does not match")
+			}
+		})
+	}
+
+}
+
+func fakeResourceRequirements() *corev1.ResourceRequirements {
```

Comment:
Maybe put this in the `testingutil` package with input based limits for future use.

Reply:
What do you think about interfacing the exported generator functions like we discussed?

Reply:
We can leave that for now as that would require a lot of unit test code changes from odo's side.